src/backend/distributed/commands/role.c

```diff
@@ -141,6 +239,100 @@ ExtractEncryptedPassword(Oid roleOid)
 }
 
 
+/*
+ * GenerateAlterRoleSetIfExistsCommandList generate a list of ALTER ROLE .. SET commands that
+ * copies a role session defaults from the pg_db_role_settings table.
+ */
+static List *
+GenerateAlterRoleSetIfExistsCommandList(HeapTuple tuple, TupleDesc
+										DbRoleSettingDescription)
+{
+	AlterRoleSetStmt *stmt = makeNode(AlterRoleSetStmt);
+	const char *currentDatabaseName = CurrentDatabaseName();
+	List *commandList = NIL;
+	bool isnull = false;
+
+	const char *databaseName =
+		GetDatabaseNameFromDbRoleSetting(tuple, DbRoleSettingDescription);
+
+	/*
+	 * session defaults for databases other than the current one are skipped
+	 */
+	if (databaseName != NULL &&
+		pg_strcasecmp(databaseName, currentDatabaseName) != 0)
+	{
+		ereport(NOTICE, (errmsg("Citus partially supports ALTER ROLE .. IN DATABASE"
+								" .. SET  for distributed databases"),
+						 errdetail("Citus propagates the session defaults that affect "
+								   "current database"),
+						 errhint("You can manually change attributes of roles "
+								 "on workers")));
+		return NULL;
+	}
+
+	if (databaseName != NULL)
+	{
+		stmt->database = pstrdup(databaseName);
+	}
+
+	const char *roleName = GetRoleNameFromDbRoleSetting(tuple, DbRoleSettingDescription);
+
+	/*
+	 * default roles are skipped, because reserved roles
+	 * cannot be altered.
+	 */
+	if (roleName != NULL && IsReservedName(roleName))
+	{
+		return NULL;
+	}
+
+	if (roleName != NULL)
+	{
+		stmt->role = makeNode(RoleSpec);
+		stmt->role->location = -1;
+		stmt->role->roletype = ROLESPEC_CSTRING;
+		stmt->role->rolename = pstrdup(roleName);
+	}
+
+	Datum setconfig = heap_getattr(tuple, Anum_pg_db_role_setting_setconfig,
+								   DbRoleSettingDescription, &isnull);
+	ArrayType *array = DatumGetArrayTypeP(setconfig);
+	int i;
+
+	for (i = 1; i <= ARR_DIMS(array)[0]; i++)
+	{
+		Datum d = array_ref(array, 1, &i,
+							-1 /* varlenarray */,
+							-1 /* TEXT's typlen */,
+							false /* TEXT's typbyval */,
+							'i' /* TEXT's typalign */,
+							&isnull);
+		if (isnull)
+		{
+			continue;
+		}
+		char *configItem = TextDatumGetCString(d);
+
+		char *pos = strchr(configItem, '=');
+		if (pos == NULL)
+		{
+			continue;
+		}
+		*pos++ = '\0';
+
+		stmt->setstmt = makeNode(VariableSetStmt);
+		stmt->setstmt->kind = VAR_SET_VALUE;
+		stmt->setstmt->name = pstrdup(configItem);
+
+		/* TODO add support for const values that are not strings */
```

Comment:
I am going to add support for Constants that are integers or floats here.

It is a bit tricky to parse a string and understand the correct data type, and value.

Similar problems were much easier to solve before we introduced `distributed_object_ops.c` where we directly wrote to a buffer without creating an `Alter*SetStmt` object

An earlier and easier solution for `AlterTableSetStmt` can be found at:
https://github.com/citusdata/citus/blob/5d601bb45a43cdc7711985fd51f670693b23240c/src/backend/distributed/deparser/citus_ruleutils.c#L1265-L1280

Reply:
I implemented the needed parsing logic in https://github.com/citusdata/citus/pull/3510/commits/87d9a10c9d7fc41d18a0f7937dd88e9e454f2731

@thanodnl can you check the changes in this commit, and comment on my approach?